setup.py

```diff
@@ -0,0 +1,201 @@
+SRC_DIR = os.path.join(TOP_DIR, 'caffe2')
+PROTOC = find_executable('protoc')
+
+install_requires = set()
+setup_requires = set()
+test_requires = set()
+
+################################################################################
+# Pre Check
+################################################################################
+
+assert find_executable('cmake'), 'Could not find "cmake" executable!'
+assert find_executable('make'), 'Could not find "make" executable!'
+
+################################################################################
+# Version
+################################################################################
+
+try:
+    git_version = subprocess.check_output(['git', 'rev-parse', 'HEAD'],
+                                          cwd=TOP_DIR).decode('ascii').strip()
+except subprocess.CalledProcessError:
+    git_version = None
+
+with open(os.path.join(TOP_DIR, 'VERSION_NUMBER')) as version_file:
+    VersionInfo = namedtuple('VersionInfo', ['version', 'git_version'])(
+        version=version_file.read().strip(),
+        git_version=git_version
+    )
+
+################################################################################
+# Customized commands
+################################################################################
+
+class Caffe2Command(setuptools.Command):
+    user_options = []
+
+    def initialize_options(self):
+        pass
+
+    def finalize_options(self):
+        pass
+
+
+class create_version(Caffe2Command):
+    def run(self):
+        with open(os.path.join(SRC_DIR, 'version.py'), 'w') as f:
+            f.write(dedent('''
+            version = '{version}'
+            git_version = '{git_version}'
+            '''.format(**dict(VersionInfo._asdict()))))
+
+
+class build_py(setuptools.command.build_py.build_py):
+    def run(self):
+        self.run_command('create_version')
+        setuptools.command.build_py.build_py.run(self)
+
+
+class develop(setuptools.command.develop.develop):
+    def run(self):
+        raise RuntimeError('develop mode is not supported!')
+
+
+class build_ext(setuptools.command.build_ext.build_ext):
+    def _build_with_cmake(self):
+        build_temp = os.path.realpath(self.build_temp)
+        build_lib = os.path.realpath(self.build_lib)
+
+        install_dir = os.environ.get('CMAKE_INSTALL_DIR',
+                                     os.path.join(build_temp, 'cmake_install'))
+        build_dir = os.environ.get('CMAKE_BUILD_DIR',
+                                   os.path.join(build_temp, 'cmake_build'))
+        if not os.path.exists(build_dir):
+            self.compiler.mkpath(build_dir)
+
+        py_exe = sys.executable
+        py_inc = sysconfig.get_python_inc()
+        cmake_args = shlex.split(os.environ.get('CMAKE_ARGS', ''))
+        log.info('CMAKE_ARGS: {}'.format(cmake_args))
+
+        cwd = os.getcwd()
+        os.chdir(build_dir)
+        self.compiler.spawn([
+            'cmake',
+            '-DBUILD_SHARED_LIBS=OFF',
```

Comment:
I suppose this ensures you get 1 big Python extension .so. Is there a way to use shared libs?

Reply:
Yep, this will create one big giant python extension.
I started with using shared libs, but it's not easy (yep there is a way, will describe later) to get it work. The problem is at the build stage, we don't know where will the final installation dir (setuptools installation dir, not the cmake installation dir) be, so we can not use an absolute rpath. Note since we want to create wheels, we can not put these cmake steps to the install* commands. The possible way to make it work is using relative rpath, but this relative value will be different between pybind_state.so and the so in $cmake_install_dir/lib directory (imagine pybind_state.so -> libcaffe2.so -> libxxx.so, the first one is at caffe2/python directory, and the latter two are in lib/ directory), so we need to do patchelf for individual so outside of cmake.
But regardless, I think it's better to do static libs/one giant so when doing deployment, it makes the deployment story much easier.